File: server/internal/wealth/yahoo_provider.go
package wealth

import (
	"cmp"
	"context"
	"fmt"
	"time"

	apiClients "tallyo/internal/clients"
	"tallyo/internal/graph/model"

	"github.com/samber/lo"
)

const (
	yahooCacheTTL = 30 * time.Minute
	// yahooPriceCacheMaxEntries bounds ticker/date keys across price refreshes.
	yahooPriceCacheMaxEntries = 256
)

type yahooCachedPrice struct {
	price  float64
	expiry time.Time
}

func (p *YahooPriceProvider) PriceAt(ctx context.Context, asset *model.Asset, date time.Time) (float64, error) {
	if asset == nil {
		return 0, fmt.Errorf("asset is nil")
	}
	if asset.AssetType == model.AssetTypeCurrency && asset.Identifier == "USD" {
		return 1, nil
	}
	if asset.AssetType == model.AssetTypeCurrency {
		return p.fallback(asset), nil
	}
	if asset.PriceConnectivity == model.ConnectivityStatusIgnore || asset.PriceConnectivity == model.ConnectivityStatusNotFound {
		return p.fallback(asset), nil
	}

	ticker := effectiveTicker(asset)
	if apiClients.IsSyntheticYahooTicker(ticker) {
		return p.fallback(asset), nil
	}
	dateStr := date.UTC().Format("2006-01-02")
	cacheKey := ticker + "|" + dateStr
	if price, ok := p.cachedPrice(cacheKey); ok {
		return adjustedPrice(price, asset), nil
	}

	price, priceAt, err := p.Client.FetchPrice(ctx, ticker, date)
	if err != nil {
		if apiClients.IsNotFound(err) {
			p.updatePriceConnectivity(ctx, asset, model.ConnectivityStatusNotFound)
		}
		p.Log.Warn("yahoo price fetch failed, using last known", "ticker", ticker, "error", err)
		return p.fallback(asset), nil
	}

	p.cachePrice(cacheKey, price)
	price = adjustedPrice(price, asset)
	p.updatePriceConnectivity(ctx, asset, model.ConnectivityStatusHealthy)
	if p.isCurrentPriceDate(date) && asset.ID.Int64() != 0 {
		p.writeAssetPrice(ctx, asset.ID.Int64(), price, priceAt)
	}
	return price, nil
}

func (p *YahooPriceProvider) cachedPrice(key string) (float64, bool) {
	p.cacheMu.Lock()
	defer p.cacheMu.Unlock()
	cached, ok := p.cache[key]
	if !ok || !p.now().Before(cached.expiry) {
		delete(p.cache, key)
		return 0, false
	}
	return cached.price, true
}

func (p *YahooPriceProvider) cachePrice(key string, price float64) {
	p.cacheMu.Lock()
	defer p.cacheMu.Unlock()
	if p.cache == nil {
		p.cache = make(map[string]yahooCachedPrice)
	}
	if _, exists := p.cache[key]; !exists && len(p.cache) == yahooPriceCacheMaxEntries {
		clear(p.cache)
	}
	p.cache[key] = yahooCachedPrice{price: price, expiry: p.now().Add(yahooCacheTTL)}
}

// ValidateTicker reports whether the ticker is known to Yahoo. It returns an
// error only when Yahoo definitively reports the symbol as not found; transient
// failures (network, rate limiting) return nil so they don't block asset edits.
func (p *YahooPriceProvider) ValidateTicker(ctx context.Context, ticker string) error {
	if apiClients.IsSyntheticYahooTicker(ticker) {
		return nil
	}
	if _, _, err := p.Client.FetchPrice(ctx, ticker, time.Now()); err != nil && apiClients.IsNotFound(err) {
		return fmt.Errorf("ticker %q not found", ticker)
	}
	return nil
}

func effectiveTicker(asset *model.Asset) string {
	if lo.FromPtr(asset.TrackingTicker) != "" {
		return *asset.TrackingTicker
	}
	return asset.Identifier
}

func adjustedPrice(price float64, asset *model.Asset) float64 {
	multiplier := cmp.Or(asset.TrackingMultiplier, 1)
	return price * multiplier
}

func (p *YahooPriceProvider) writeAssetPrice(ctx context.Context, assetID int64, price float64, priceAt time.Time) {
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	if err := p.Store.UpdateAssetPrice(ctx, assetID, price, priceAt); err != nil {
		p.Log.Warn("yahoo: failed to write back asset price", "asset_id", assetID, "error", err)
	}
}

func (p *YahooPriceProvider) updatePriceConnectivity(ctx context.Context, asset *model.Asset, status model.ConnectivityStatus) {
	assetID := asset.ID.Int64()
	if assetID == 0 {
		return
	}
	if asset.PriceConnectivity == status {
		return
	}
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	if err := p.Store.UpdateAssetPriceConnectivity(ctx, assetID, status); err != nil {
		p.Log.Warn("yahoo: failed to update price connectivity", "asset_id", assetID, "status", status, "error", err)
	}
}

func (p *YahooPriceProvider) now() time.Time {
	return p.Now()
}

func (p *YahooPriceProvider) isCurrentPriceDate(date time.Time) bool {
	today := p.now().UTC().Truncate(24 * time.Hour)
	return !date.UTC().Truncate(24 * time.Hour).Before(today)
}

func (p *YahooPriceProvider) fallback(asset *model.Asset) float64 {
	return lo.FromPtr(asset.CurrentPrice)
}

File: server-rs/src/wealth/yahoo_provider.rs
use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
    time::Duration,
};

use anyhow::Result;
use chrono::{DateTime, Utc};
use sqlx::SqlitePool;

use crate::{
    clients::yahoo::{self, Yahoo},
    schema::{AssetType, ConnectivityStatus},
    wealth::{Asset, store},
};

const CACHE_TTL: Duration = Duration::from_secs(30 * 60);
const CACHE_MAX_ENTRIES: usize = 256;

#[derive(Clone, Copy)]
struct CachedPrice {
    price: f64,
    expires_at: std::time::Instant,
}

#[derive(Clone)]
pub struct YahooPriceProvider {
    pool: SqlitePool,
    client: Yahoo,
    cache: Arc<Mutex<HashMap<String, CachedPrice>>>,
}

impl YahooPriceProvider {
    pub fn new(pool: SqlitePool) -> Result<Self> {
        Ok(Self::with_client(pool, Yahoo::new()?))
    }

    pub fn with_client(pool: SqlitePool, client: Yahoo) -> Self {
        Self {
            pool,
            client,
            cache: Arc::default(),
        }
    }

    pub async fn price_at(&self, asset: &Asset, date: DateTime<Utc>) -> Result<f64> {
        if asset.asset_type == AssetType::Currency && asset.identifier == "USD" {
            return Ok(1.0);
        }
        if asset.asset_type == AssetType::Currency
            || matches!(
                asset.price_connectivity,
                ConnectivityStatus::Ignore | ConnectivityStatus::NotFound
            )
        {
            return Ok(fallback(asset));
        }
        let ticker = effective_ticker(asset);
        if yahoo::is_synthetic_yahoo_ticker(ticker) {
            return Ok(fallback(asset));
        }
        let key = format!("{ticker}|{}", date.format("%F"));
        if let Some(price) = self.cached_price(&key) {
            return Ok(adjusted_price(price, asset));
        }
        match self.client.fetch_price(ticker, date).await {
            Ok((price, price_at)) => {
                self.cache_price(key, price);
                let price = adjusted_price(price, asset);
                self.update_price_connectivity(asset, ConnectivityStatus::Healthy).await;
                if is_current_price_date(date) && asset.id != 0 {
                    self.write_asset_price(asset.id, price, price_at).await;
                }
                Ok(price)
            }
            Err(error) => {
                if yahoo::is_not_found(&error) {
                    self.update_price_connectivity(asset, ConnectivityStatus::NotFound)
                        .await;
                }
                Ok(fallback(asset))
            }
        }
    }

    pub async fn validate_ticker(&self, ticker: &str) -> Result<()> {
        if yahoo::is_synthetic_yahoo_ticker(ticker) {
            return Ok(());
        }
        if let Err(error) = self.client.fetch_price(ticker, Utc::now()).await
            && yahoo::is_not_found(&error)
        {
            anyhow::bail!("ticker {ticker:?} not found");
        }
        Ok(())
    }

    fn cached_price(&self, key: &str) -> Option<f64> {
        let mut cache = self.cache.lock().expect("Yahoo cache mutex must not be poisoned");
        match cache
            .get(key)
            .filter(|cached| std::time::Instant::now() < cached.expires_at)
        {
            Some(cached) => Some(cached.price),
            None => {
                cache.remove(key);
                None
            }
        }
    }

    fn cache_price(&self, key: String, price: f64) {
        let mut cache = self.cache.lock().expect("Yahoo cache mutex must not be poisoned");
        if !cache.contains_key(&key) && cache.len() == CACHE_MAX_ENTRIES {
            cache.clear();
        }
        cache.insert(
            key,
            CachedPrice {
                price,
                expires_at: std::time::Instant::now() + CACHE_TTL,
            },
        );
    }

    async fn write_asset_price(&self, asset_id: i64, price: f64, price_at: DateTime<Utc>) {
        let _ = tokio::time::timeout(
            Duration::from_secs(5),
            store::update_asset_price(&self.pool, asset_id, price, price_at),
        )
        .await;
    }

    async fn update_price_connectivity(&self, asset: &Asset, status: ConnectivityStatus) {
        if asset.id == 0 || asset.price_connectivity == status {
            return;
        }
        let _ = tokio::time::timeout(
            Duration::from_secs(5),
            store::update_asset_price_connectivity(&self.pool, asset.id, status),
        )
        .await;
    }
}

fn effective_ticker(asset: &Asset) -> &str {
    asset.tracking_ticker.as_deref().unwrap_or(&asset.identifier)
}

fn adjusted_price(price: f64, asset: &Asset) -> f64 {
    price * if asset.tracking_multiplier == 0.0 { 1.0 } else { asset.tracking_multiplier }
}

fn fallback(asset: &Asset) -> f64 {
    asset.current_price.unwrap_or_default()
}

fn is_current_price_date(date: DateTime<Utc>) -> bool {
    date.date_naive() >= Utc::now().date_naive()
}

#[cfg(test)]
mod tests {
    use chrono::{DateTime, Duration, Utc};
    use wiremock::{
        Mock, MockServer, ResponseTemplate,
        matchers::{method, path_regex, query_param},
    };

    use super::YahooPriceProvider;
    use crate::{
        clients::yahoo::Yahoo,
        schema::{AssetClassifier, AssetType, ConnectivityStatus},
        wealth::Asset,
    };

    #[tokio::test]
    async fn prices_usd_and_skips_unpriceable_assets_without_network() {
        let server = MockServer::start().await;
        let provider = YahooPriceProvider::with_client(
            crate::database::dbtest::open().await.unwrap(),
            Yahoo::with_base_url(server.uri()).unwrap(),
        );

        assert_eq!(
            provider
                .price_at(&asset(AssetType::Currency, "USD"), Utc::now())
                .await
                .unwrap(),
            1.0
        );
        assert_eq!(
            provider
                .price_at(
                    &Asset {
                        current_price: Some(42.0),
                        price_connectivity: ConnectivityStatus::Ignore,
                        ..asset(AssetType::Security, "AAPL")
                    },
                    Utc::now(),
                )
                .await
                .unwrap(),
            42.0
        );
        assert_eq!(
            provider
                .price_at(&asset(AssetType::Security, "PLAID:cash"), Utc::now())
                .await
                .unwrap(),
            0.0
        );
    }

    #[tokio::test]
    async fn caches_adjusted_prices_and_uses_the_prior_historical_close() {
        let server = MockServer::start().await;
        let date = Utc::now() - Duration::days(2);
        chart_mock(date).expect(1).mount(&server).await;
        let provider = YahooPriceProvider::with_client(
            crate::database::dbtest::open().await.unwrap(),
            Yahoo::with_base_url(server.uri()).unwrap(),
        );
        let asset = Asset {
            tracking_multiplier: 2.0,
            ..asset(AssetType::Security, "AAPL")
        };

        assert_eq!(provider.price_at(&asset, date).await.unwrap(), 22.0);
        assert_eq!(provider.price_at(&asset, date).await.unwrap(), 22.0);
        server.verify().await;
    }

    #[tokio::test]
    async fn validates_only_definitively_missing_tickers() {
        let server = MockServer::start().await;
        Mock::given(method("GET"))
            .respond_with(ResponseTemplate::new(404))
            .mount(&server)
            .await;
        let provider = YahooPriceProvider::with_client(
            crate::database::dbtest::open().await.unwrap(),
            Yahoo::with_base_url(server.uri()).unwrap(),
        );

        assert_eq!(
            provider.validate_ticker("MISSING").await.unwrap_err().to_string(),
            "ticker \"MISSING\" not found"
        );
        assert!(provider.validate_ticker("PLAID:cash").await.is_ok());
    }

    #[tokio::test]
    async fn clones_share_one_price_cache() {
        let server = MockServer::start().await;
        let date = Utc::now() - Duration::days(2);
        chart_mock(date).expect(1).mount(&server).await;
        let provider = YahooPriceProvider::with_client(
            crate::database::dbtest::open().await.unwrap(),
            Yahoo::with_base_url(server.uri()).unwrap(),
        );
        let clone = provider.clone();
        let asset = asset(AssetType::Security, "AAPL");

        assert_eq!(provider.price_at(&asset, date).await.unwrap(), 11.0);
        assert_eq!(clone.price_at(&asset, date).await.unwrap(), 11.0);
        server.verify().await;
    }

    fn chart_mock(date: DateTime<Utc>) -> Mock {
        Mock::given(method("GET"))
            .and(path_regex("/v8/finance/chart/.*"))
            .and(query_param("interval", "1d"))
            .respond_with(ResponseTemplate::new(200).set_body_string(format!(
                r#"{{"chart":{{"result":[{{"indicators":{{"quote":[{{"close":[11.0,12.0]}}]}},"timestamp":[{},{}]}}]}}}}"#,
                (date - Duration::days(1)).timestamp(),
                (date + Duration::days(1)).timestamp(),
            )))
    }

    fn asset(asset_type: AssetType, identifier: &str) -> Asset {
        Asset {
            id: 0,
            asset_type,
            identifier: identifier.to_owned(),
            name: None,
            classifier: AssetClassifier::Public,
            current_price: None,
            forced_usd_price: None,
            tracking_ticker: None,
            tracking_multiplier: 1.0,
            price_connectivity: ConnectivityStatus::Healthy,
            investment_connectivity: ConnectivityStatus::Healthy,
        }
    }
}
